# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/libwebrtcglue/AudioConduit.h
# Commit: 3fb02a012faa
# Full Hash: 3fb02a012faadb59c6a4ca42bfb50bcbb1037d93
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 5
# Description:
#   Bug 1654112 - Route packets with MediaEvents, and clean up conduits' cross-thread syncing. r=bwc,ng
#   
#   This patch solves some issues with the SSRC switching that can be triggered by
#   both signaling and incoming packets, on different threads. But mainly, because
#   that setup is inherently complex, this patch greatly simplifies the SSRC
# ==============================================================================

diff -r 336f58a54bd8 -r 3fb02a012faa dom/media/webrtc/libwebrtcglue/AudioConduit.h
--- a/dom/media/webrtc/libwebrtcglue/AudioConduit.h	Wed Aug 25 20:28:12 2021 +0200
+++ b/dom/media/webrtc/libwebrtcglue/AudioConduit.h	Thu Sep 09 17:42:30 2021 +0200
@@ -13,7 +13,6 @@
 
 #include "MediaConduitInterface.h"
 #include "common/MediaEngineWrapper.h"
-#include "RtpPacketQueue.h"
 
 /**
  * This file hosts several structures identifying different aspects of a RTP
@@ -28,24 +27,62 @@
  *  - media-source and target to external transport
  */
 class WebrtcAudioConduit : public AudioSessionConduit,
-                           public webrtc::RtcpEventObserver,
-                           public webrtc::Transport {
+                           public webrtc::RtcpEventObserver {
  public:
-  /**
-   * APIs used by the registered external transport to this Conduit to
-   * feed in received RTP Frames to the Receiver for decoding.
-   */
-  void ReceivedRTPPacket(const uint8_t* data, int len,
-                         webrtc::RTPHeader& header) override;
+  void OnRtpReceived(MediaPacket&& aPacket, webrtc::RTPHeader&& aHeader);
+  void OnRtcpReceived(MediaPacket&& aPacket);
 
   void OnRtcpBye() override;
   void OnRtcpTimeout() override;
 
-  /**
-   * APIs used by the registered external transport to this Conduit to
-   * feed in received RTCP Frames to the Receiver for decoding.
-   */
-  void ReceivedRTCPPacket(const uint8_t* data, int len) override;
+  void SetTransportActive(bool aActive) override {
+    mTransportActive = aActive;
+    if (!aActive) {
+      mReceiverRtpEventListener.DisconnectIfExists();
+      mReceiverRtcpEventListener.DisconnectIfExists();
+      mSenderRtcpEventListener.DisconnectIfExists();
+    }
+  }
+  MediaEventSourceExc<MediaPacket>& SenderRtpSendEvent() override {
+    return mSenderRtpSendEvent;
+  }
+  MediaEventSourceExc<MediaPacket>& SenderRtcpSendEvent() override {
+    return mSenderRtcpSendEvent;
+  }
+  MediaEventSourceExc<MediaPacket>& ReceiverRtcpSendEvent() override {
+    return mReceiverRtcpSendEvent;
+  }
+  void ConnectReceiverRtpEvent(
+      MediaEventSourceExc<MediaPacket, webrtc::RTPHeader>& aEvent) override {
+    // Hold a strong-ref to `this` for safety, since we'll be disconnecting
+    // off-target.
+    mReceiverRtpEventListener = aEvent.Connect(
+        mCallThread, [this, self = RefPtr<WebrtcAudioConduit>(this)](
+                         MediaPacket aPacket, webrtc::RTPHeader aHeader) {
+          OnRtpReceived(std::move(aPacket), std::move(aHeader));
+        });
+  }
+  void ConnectReceiverRtcpEvent(
+      MediaEventSourceExc<MediaPacket>& aEvent) override {
+    // Hold a strong-ref to `this` for safety, since we'll be disconnecting
+    // off-target.
+    mReceiverRtcpEventListener = aEvent.Connect(
+        mCallThread,
+        [this, self = RefPtr<WebrtcAudioConduit>(this)](MediaPacket aPacket) {
+          OnRtcpReceived(std::move(aPacket));
+        });
+  }
+  void ConnectSenderRtcpEvent(
+      MediaEventSourceExc<MediaPacket>& aEvent) override {
+    // Hold a strong-ref to `this` for safety, since we'll be disconnecting
+    // off-target.
+    mSenderRtcpEventListener = aEvent.Connect(
+        mCallThread,
+        [this, self = RefPtr<WebrtcAudioConduit>(this)](MediaPacket aPacket) {
+          OnRtcpReceived(std::move(aPacket));
+        });
+  }
+
   Maybe<DOMHighResTimeStamp> LastRtcpReceived() const override;
   DOMHighResTimeStamp GetNow() const override;
 
@@ -55,17 +92,6 @@
   MediaConduitErrorCode StartReceivingLocked();
 
   /**
-   * Register External Transport to this Conduit. RTP and RTCP frames from the
-   * VoiceEngine shall be passed to the registered transport for transporting
-   * externally.
-   */
-  MediaConduitErrorCode SetTransmitterTransport(
-      RefPtr<TransportInterface> aTransport) override;
-
-  MediaConduitErrorCode SetReceiverTransport(
-      RefPtr<TransportInterface> aTransport) override;
-
-  /**
    * Function to deliver externally captured audio sample for encoding and
    * transport
    * @param frame [in]: AudioFrame in upstream's format for forwarding to the
@@ -95,16 +121,10 @@
   MediaConduitErrorCode GetAudioFrame(int32_t samplingFreqHz,
                                       webrtc::AudioFrame* frame) override;
 
-  /**
-   * Webrtc transport implementation to send and receive RTP packet.
-   */
-  bool SendRtp(const uint8_t* data, size_t len,
-               const webrtc::PacketOptions& options) override;
-
-  /**
-   * Webrtc transport implementation to send and receive RTCP packet.
-   */
-  bool SendRtcp(const uint8_t* data, size_t len) override;
+  bool SendRtp(const uint8_t* aData, size_t aLength,
+               const webrtc::PacketOptions& aOptions) override;
+  bool SendSenderRtcp(const uint8_t* aData, size_t aLength) override;
+  bool SendReceiverRtcp(const uint8_t* aData, size_t aLength) override;
 
   bool HasCodecPluginID(uint64_t aPluginID) const override { return false; }
 
@@ -184,18 +204,15 @@
   MediaConduitErrorCode CreateRecvStream();
   void DeleteRecvStream();
 
-  mozilla::ReentrantMonitor mTransportMonitor;
-
-  // Accessed on any thread under mTransportMonitor.
-  RefPtr<TransportInterface> mTransmitterTransport;
-
-  // Accessed on any thread under mTransportMonitor.
-  RefPtr<TransportInterface> mReceiverTransport;
-
   // Const so can be accessed on any thread. Most methods are called on the Call
   // thread.
   const RefPtr<WebrtcCallWrapper> mCall;
 
+  // Set up in the ctor and then not touched. Called through by the streams on
+  // any thread.
+  WebrtcSendTransport mSendTransport;
+  WebrtcReceiveTransport mRecvTransport;
+
   // Accessed only on the Call thread.
   webrtc::AudioReceiveStream::Config mRecvStreamConfig;
 
@@ -210,12 +227,6 @@
   // Call thread.
   webrtc::AudioSendStream* mSendStream;
 
-  // accessed on creation, and when receiving packets
-  Atomic<uint32_t> mRecvSSRC;  // this can change during a stream!
-
-  // Accessed only on mStsThread.
-  RtpPacketQueue mRtpPacketQueue;
-
   // If true => mSendStream started and not stopped
   // Written only on the Call thread. Guarded by mLock, except for reads on the
   // Call thread.
@@ -271,12 +282,21 @@
   // Accessed from mStsThread. Last successfully polled RTT
   Maybe<DOMHighResTimeStamp> mRttSec;
 
-  // Accessed only on mStsThread
+  // Call thread.
   Maybe<DOMHighResTimeStamp> mLastRtcpReceived;
 
   // Thread safe
+  Atomic<bool> mTransportActive = Atomic<bool>(false);
   MediaEventProducer<void> mRtcpByeEvent;
   MediaEventProducer<void> mRtcpTimeoutEvent;
+  MediaEventProducerExc<MediaPacket> mSenderRtpSendEvent;
+  MediaEventProducerExc<MediaPacket> mSenderRtcpSendEvent;
+  MediaEventProducerExc<MediaPacket> mReceiverRtcpSendEvent;
+
+  // Assigned and revoked on mStsThread. Listeners for receiving packets.
+  MediaEventListener mSenderRtcpEventListener;    // Rtp-transmitting pipeline
+  MediaEventListener mReceiverRtcpEventListener;  // Rtp-receiving pipeline
+  MediaEventListener mReceiverRtpEventListener;   // Rtp-receiving pipeline
 };
 
 }  // namespace mozilla